### PR TITLE
Add custom function findMax()

### PR DESCRIPTION
**Pull request by LAV45**

### CODE REVIEW

**Comment by arogachev** [comment]
Why user need to specify it? I think this method should be internal.

**Comment by himiklab** [comment]
Give an example, please.

**Comment by arogachev** [comment]
Do you ask @LAV45 or me?

**Comment by himiklab** [comment]
You.

**Comment by arogachev** [comment]
Seems like I overlooked the point of that.

This is kind of "sortable scope", but `max` is definetely redundant here.

File: SortableGridBehavior.php [hunk 1]
```diff
@@ -65,13 +78,18 @@ public function gridSort($items)
...
-        /** @var ActiveRecord $model */
-        $model = $this->owner;
-        if (!$model->hasAttribute($this->sortableAttribute)) {
+        if (!$this->owner->hasAttribute($this->sortableAttribute)) {
             throw new InvalidConfigException("Invalid sortable attribute `{$this->sortableAttribute}`.");
         }
+        $this->owner->{$this->sortableAttribute} = $this->findMax() + 1;
+    }
 
-        $maxOrder = $model->find()->max($model->tableName() . '.' . $this->sortableAttribute);
```
Comment (LAV45) [inline SortableGridBehavior.php]: In principle, I have nothing to add, but only extended the current functionality and made the string into a separate method.

**Comment by LAV45** [comment]
## Example
One product can have multiple images

## Table Image

id | product_id | sort_order
---------- | ---------- | ----------
1 | 1 | 1
2 | 1 | 2
3 | 1 | 3
4 | 2 | 4
5 | 2 | 5

When you add new pictures `sort_order` increases to max + 1
`sort_order` almost copies the `id` field

When sorting images in one product value fields sort_order just swapped while sort_order field is mandatory and must be filled

id | product_id | sort_order
---------- | ---------- | ----------
1 | 1 | 3
2 | 1 | 2
3 | 1 | 1
4 | 2 | 5
5 | 2 | 4

because the items are sorted only within their product then there is no need to set a field sort_order huge value. In DB (mysql) for the field sort_order you can use tinyint (1)

id | product_id | sort_order
---------- | ---------- | ----------
1 | 1 | 3
2 | 1 | 2
3 | 1 | 1
4 | 2 | 2
5 | 2 | 1

I made a string with maximum values sort_order in a separate method findMax, so you can add your own rules for getting the max sort_order

**Comment by arogachev** [comment]
I think user should care only about setting sortable scope condition, and not about finding `max`. And you kind of mixing logic of setting sortable scope and finding `max`.

**Comment by arogachev** [comment]
Also you are using different ways of storing order for different products.

**Comment by LAV45** [comment]
That's why I gave the example of code in comments, because it is not mandatory. findMax you can customize as needed.

```php
use himiklab\sortablegrid\SortableGridBehavior;

public function behaviors()
{
    return [
        [
            'class' => SortableGridBehavior::className(),
//            'sortableAttribute' => 'sortOrder'
//            'findMax' => function() {
//                return self::find()
//                    ->where(['product_id' => $this->product_id])
//                    ->max('sort_order');
//            },
        ],
    ];
}
```

**Comment by arogachev** [comment]
Still doesn't make sense for me.

**Comment by LAV45** [comment]
:+1: cool, offer your

**Comment by arogachev** [comment]
You can see how I handle this in my similar extension.
https://github.com/arogachev/yii2-sortable

**Comment by arogachev** [comment]
So the user shouldn't think about implementation details, finding maximums, etc.

**Pull request closed without merging**